packages/front-end/components/Dimensions/DimensionForm.tsx

```diff
@@ -31,101 +33,121 @@ const DimensionForm: FC<{
       owner: current.owner || "",
     },
   });
+  const [sqlOpen, setSqlOpen] = useState(false);
 
   const datasource = form.watch("datasource");
   const userIdType = form.watch("userIdType");
 
   const dsObj = getDatasourceById(datasource);
+
   const dsProps = dsObj?.properties;
-  const sql = dsProps?.queryLanguage === "sql";
+  const supportsSQL = dsProps?.queryLanguage === "sql";
+
+  const sql = form.watch("sql");
 
   const requiredColumns = useMemo(() => {
     return new Set([userIdType, "value"]);
   }, [userIdType]);
 
   return (
-    <Modal
-      close={close}
-      open={true}
-      header={current.id ? "Edit Dimension" : "New Dimension"}
-      submit={form.handleSubmit(async (value) => {
-        if (sql) {
-          validateSQL(value.sql, [value.userIdType, "value"]);
-        }
-
-        await apiCall(
-          current.id ? `/dimensions/${current.id}` : `/dimensions`,
-          {
-            method: current.id ? "PUT" : "POST",
-            body: JSON.stringify(value),
-          }
-        );
-        mutateDefinitions();
-      })}
-    >
-      <Field label="Name" required {...form.register("name")} />
-      <Field
-        label="Owner"
-        options={memberUsernameOptions}
-        comboBox
-        {...form.register("owner")}
-      />
-      <SelectField
-        label="Data Source"
-        required
-        value={form.watch("datasource")}
-        onChange={(v) => form.setValue("datasource", v)}
-        placeholder="Choose one..."
-        options={datasources.map((d) => ({
-          value: d.id,
-          label: `${d.name}${d.description ? ` — ${d.description}` : ""}`,
-        }))}
-        className="portal-overflow-ellipsis"
-      />
-      {dsProps.userIds && (
-        <SelectField
-          label="Identifier Type"
-          required
-          value={userIdType}
-          onChange={(v) => form.setValue("userIdType", v)}
-          options={(dsObj.settings.userIdTypes || []).map((t) => {
-            return {
-              label: t.userIdType,
-              value: t.userIdType,
-            };
-          })}
+    <>
+      {sqlOpen && datasource && (
+        <EditSqlModal
+          close={() => setSqlOpen(false)}
+          datasourceId={dsObj.id || ""}
```

Comment:
dsObj might be undefined here.  Above where you're doing `sqlOpen && datasource &&`, you can change it to `sqlOpen && dsObj &&` instead.